### PR TITLE
[addons] fix multiple usage

### PR DESCRIPTION
**Pull request by AlwinEsch**
## Description
If one addon defines several types inside, was before always only the first
possible to start, e.g. screensaver.atv4 has also a video addon inside,
but this was before not possible to start.

The view, control and start of addons where have more as one type inside are now fixed with this request.

## Screenshots (if appropriate):
Now with more correct view:
Bildschirmfoto von 2020-01-28 01-50-32

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/addons/PluginSource.cpp [hunk 1]
```diff
@@ -102,7 +102,8 @@ bool CPluginSource::IsType(TYPE type) const
        || (type == ADDON_AUDIO && Provides(AUDIO))
        || (type == ADDON_IMAGE && Provides(IMAGE))
        || (type == ADDON_GAME && Provides(GAME))
-       || (type == ADDON_EXECUTABLE && Provides(EXECUTABLE)));
+       || (type == ADDON_EXECUTABLE && Provides(EXECUTABLE))
+       || CAddon::IsType(type)); // Fallback also to other types e.g. ADDON_SCRIPT is asked
```
Comment (enen92) [inline xbmc/addons/PluginSource.cpp]: Why is this change needed?

Comment (enen92) [reply xbmc/addons/PluginSource.cpp]: It doesn't make much sense to me. A plugin is a plugin, why do you check it against `ADDON_SCRIPT` ?

Comment (AlwinEsch) [reply xbmc/addons/PluginSource.cpp]: The directory ask always for `ADDON_SCRIPT` but the scripts are also handled by `CPluginSource` where primary for `ADDON_PLUGIN`, without it becomes never a true if asked for.

File: xbmc/addons/Addon.h [hunk 2]
```diff
@@ -35,7 +35,7 @@ class CAddon : public IAddon
 
   TYPE Type() const override { return m_addonInfo->MainType(); }
   TYPE FullType() const override { return Type(); }
-  bool IsType(TYPE type) const override { return type == m_addonInfo->MainType(); }
+  bool IsType(TYPE type) const override { return m_addonInfo->IsType(type); }
```
Comment (enen92) [inline xbmc/addons/Addon.h]: Wouldn't it be easier to:

- Change `Type()` function to `MainType()` as the function name is misleading
- Create a new method `bool HasType(Type)` that would return `true` if an addon has the given type
- Just use HasType(xx) in the switch/if conditions

?

Comment (AlwinEsch) [reply xbmc/addons/Addon.h]: With `Type()`I can change, was only leaved as them to have the cpluff remove request not bigger with rename of them.

With new method `bool HasType(Type)`, what should be different to `bool IsType(TYPE type)`? 

Or you mean to leave the ` { return type == m_addonInfo->MainType(); }`, this was I think only added from me to prevent this view the 2 new types on addon Window (If I remember correctly)?

Comment (enen92) [reply xbmc/addons/Addon.h]: I just gave it a try and it does indeed fix the issue with `screensaver.atv4`. However I don't like the Scripts and Media sources showing in the addon browser.

What I mean with my suggestion was exactly that. `Type()` and `IsType()` refer to MainType only so they should be renamed to reflect what they do. `HasType` (or your new `IsType`) would have the implementation you provide in this PR. The only code block using `HasType` would be `GetScriptsAndPlugins` for now.

**Comment by enen92** [comment]
JFY: Issue with `embruary.info` still persists after this PR:

alt text

If you install the addon from the official repo, you'll find 2 icons under video addons. One of the entrypoints is a script (and should show under Program Addons)

**Comment by enen92** [comment]
Fixes https://github.com/xbmc/xbmc/issues/17125 (cross-reference)

@AlwinEsch I know you're quite busy with binary addons. I'm wondering if you could pick this up whenever possible and find a way of not presenting "Scripts" and "Media Sources" in the interface. Unfortunately without this getting fix I can't work on my screensaver.
Cheers

**Comment by AlwinEsch** [comment]
I hope to go soon back on them.

File: xbmc/addons/PluginSource.cpp [hunk 3]
```diff
@@ -80,29 +86,11 @@ CPluginSource::Content CPluginSource::Translate(const std::string &content)
...
+bool CPluginSource::HasType(TYPE type) const
 {
-  return ((type == ADDON_VIDEO && Provides(VIDEO))
-       || (type == ADDON_AUDIO && Provides(AUDIO))
-       || (type == ADDON_IMAGE && Provides(IMAGE))
-       || (type == ADDON_GAME && Provides(GAME))
-       || (type == ADDON_EXECUTABLE && Provides(EXECUTABLE)));
+  return ((type == ADDON_VIDEO && Provides(VIDEO)) || (type == ADDON_AUDIO && Provides(AUDIO)) ||
+          (type == ADDON_IMAGE && Provides(IMAGE)) || (type == ADDON_GAME && Provides(GAME)) ||
+          (type == ADDON_EXECUTABLE && Provides(EXECUTABLE)) || (type == CAddon::Type()));
```
Comment (garbear) [inline xbmc/addons/PluginSource.cpp]: Should we keep these on separate lines for readability and git history niceness?

Comment (AlwinEsch) [reply xbmc/addons/PluginSource.cpp]: Aha github is working again :smile: 
No problem hold it also for better like before was only about the guidelines. 
I change it.

**Comment by AlwinEsch** [comment]
This request is complete udpated and have mostly now the same view as on Leia, also was the test with Video and App addons OK.

Also renamed the IsType to HasType, does more match the wanted work.

Here is also (for testing) this request changes on Version with Python 2: https://github.com/AlwinEsch/kodi/tree/fix-multi-addons-python2

**Review by garbear — APPROVED** [review]

**Review by enen92 — APPROVED** [review]
I can confirm this fixes https://github.com/xbmc/xbmc/issues/17125 and also the issue affecting embruary.info. Thanks a lot!
